File: Semana 2/Sabado 6-09/opp-sample/src/main/java/acme/opp/sales/domain/model/aggregates/SalesOrderItem.java
package acme.opp.sales.domain.model.aggregates;

import acme.opp.crm.domain.model.valueobjects.ProductId;
import acme.opp.shared.domain.model.valueobjects.Money;
import lombok.Getter;
import lombok.Setter;

import java.math.BigDecimal;
import java.util.Objects;

/*
 * Represents the SalesOrderItem
 */

public class SalesOrderItem {
    private final ProductId productId;
    @Setter private int quantity;
    @Setter private Money unitPrice;

    /**
     * Constructs a SalesOrderItem with the specified product ID, quantity, and unit price.
     * @param productId the unique identifier of the product.
     * @param quantity the quantity of the product ordered, must be greater than zero.
     * @param unitPrice the quantity of the product ordered, must be greater than zero.
     * @throws IllegalArgumentException if quantity is less than or equal to zero,
     *                                  if unitprice is not positive,
     *                                  or if unitPrice does not have a valid currency.
     */
    SalesOrderItem(ProductId productId, int quantity, Money unitPrice) {
        if (quantity <= 0){
            throw new IllegalArgumentException("Quantity must be greater than zero");
        }
        if (unitPrice.amount().compareTo(BigDecimal.ZERO) <= 0){
            throw new IllegalArgumentException("Unit price must be greater than zero");
        }
        if (Objects.isNull(unitPrice.currency()) || Objects.isNull(unitPrice.currency().getCurrencyCode())
            || unitPrice.currency().getCurrencyCode().isBlank()) {
               throw new IllegalArgumentException("Unit price currency code must not be blank");
        }
        this.productId = productId;
        this.quantity = quantity;
        this.unitPrice = unitPrice;
    }

    /**
     * Calculates the total price for this sales order item
     * @return the total price as a Money object, which is the product of unit price and quantity.
     */

    public Money calculateItemAmount() {
        return unitPrice.multiply(BigDecimal.valueOf(quantity));
    }
}

File: Semana 3/Sabado 13-09/typescript-review/src/sales/domain/model/salesOrderItem.ts
import {Money} from "../../../shared/domain/model/money";
import {ProductId} from "./product-id";

/**
 * SalesOrderItem represents an item in a sales order, aggregate within the Sales bounded context.
 */
export class SalesOrderItem {
    private readonly _orderId: string;
    private readonly _itemId: string;
    private readonly _productId: ProductId;
    private readonly _quantity: number;
    private readonly _unitPrice: Money;

    /**
     * Creates a new SalesOrderItem.
     * @param orderId   - ID of the order this item belongs to
     * @param productId - ID of the product
     * @param quantity - Quantity of the product
     * @param unitPrice - Unit price of the product
     * @throws Error if quantity is less than or equal to zero
     */
    constructor(orderId: string, productId: ProductId, quantity: number, unitPrice: Money) {
        if (quantity <= 0) throw Error("Quantity must be greater than zero");
        this._orderId = orderId;
        this._itemId = crypto.randomUUID();
        this._productId = productId;
        this._quantity = quantity;
        this._unitPrice = unitPrice;
    }

    public get orderId(): string {  return this._orderId;     }

    public get itemId(): string {        return this._itemId;    }

    public get productId(): ProductId {        return this._productId;    }

    public get quantity(): number {        return this._quantity;    }

    public get unitPrice(): Money {        return this._unitPrice;    }

    /**
     * Calculates the total price for this item (unit price * quantity).
     * @returns Money representing the total price for this item
     */
    public calculateItemTotal(): Money {
        return new Money(this._unitPrice.amount * this._quantity, this._unitPrice.currency);
    }
}